Add `axis_changed` event and use it in `any_axis_changed` event

francisco-dlp opened the pull request:
### Description of the change

The `any_axis_changed` event predates the implementation of the different types of axes that HyperSpy supports, and only works for UniformDataAxis. This PR fixes this bug by implementing a new `axis_changed` event that monitor changes to any individual axes attributes in all types of axes and use it in the new implementation of `any_axis_changed`.
### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [ ] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

francisco-dlp committed:
Add axis_changed event and use it in any_axis_changed

ericpre commented on hyperspy/axes.py:
```diff
@@ -343,12 +363,19 @@ def __init__(
 
     def _index_changed(self, name, old, new):
         self.events.index_changed.trigger(obj=self, index=self.index)
+        if not self._suppress_axis_changed_trigger:
```
Why should an `axis_changed` event be trigger when the index is changed?

francisco-dlp replied:
Good point, fixed!

francisco-dlp committed:
Fix tests

francisco-dlp committed:
Improve coverage

francisco-dlp committed:
Fix formatting issues in upcoming_changes entries and add issue number

ericpre reviewed (commented):
I had a quick look and this look good, except one question.

I will still need to review the added tests.

francisco-dlp committed:
Further increase coverage

francisco-dlp committed:
Fix _get_signal_signal to return non-lazy signal with numpy data

When called on a lazy signal with data=None, _get_signal_signal
previously returned a lazy signal class (e.g. LazySignal1D) with
numpy data, which is inconsistent with _get_navigation_signal that
returns a non-lazy signal in the same situation.

This fix makes _get_signal_signal use explicit non-lazy base classes
(Signal1D, Signal2D, BaseSignal) instead of self.__class__, matching
the behavior of _get_navigation_signal. The signal is still converted
to lazy when dask data is provided.

Fixes inconsistency where _lazy=True but data was numpy array.

francisco-dlp committed:
Merge remote-tracking branch 'origin/RELEASE_next_patch' into fix-lazy-signal-consistency

francisco-dlp committed:
Merge pull request #3620 from francisco-dlp/fix-lazy-signal-consistency

Fix _get_signal_signal to return non-lazy signal with numpy data

francisco-dlp committed:
Merge remote-tracking branch 'origin/RELEASE_next_patch' into NEW_axis_changed_event

# Conflicts:
#	hyperspy/axes.py

francisco-dlp committed:
Remove axis_changed trigger from _index_changed and _value_changed

axis_changed should only fire when the axis definition changes (scale,
offset, name, units, parameters), not on navigation events (index/value
changes). The on_trait_change registrations for _axis_changed and the
update_axis() methods already cover all definition changes.

Thanks to @ericpre for raising this in the PR review.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Merge remote-tracking branch 'origin/RELEASE_next_minor' into NEW_axis_changed_event

# Conflicts:
#	hyperspy/tests/axes/test_axes_manager.py
